### DISCUSSION TITLE
working with pivot tables

### DISCUSSION THREAD

**Discussion by andyford** [adonisjs/core#1276] [Help] [answered]
does anyone have a working example of posts/tags (or any entity and tags really)?

I'm using Adonis 4.1 and I  have a posts table, a tags table and a posts_tags pivot table but I really can't  tell from the docs what the idiomatic adonis way should be to work with all of them elegantly... it seems .attach() and .detach() probably should be involved but I've not seen any clear holistic working examples of using pivot tables in adonis

I think there must be an easier way, but right now, I have a bunch of horrible manual code which works but is quite a mess...

```
const postId = 123;
const incomingTags = ['foo', 'bar', 'baz'];

// get any pre-existing tags from the list of incoming tags
// then put their string titles into an array (existingTagTitles)
const existingTags = await Database.from('tags').whereIn('title', incomingTags);
const existingTagTitles = [];
existingTags.forEach((tag) => existingTagTitles.push(tag.title));

// we only want to save the tags that weren't previously saved
// so check incoming tags against previously saved tags and create array of new tags to save
const newTagsToSave = [];
incomingTags.forEach((tag) => {
    if (!existingTagTitles.includes(tag)) {
        newTagsToSave.push({ title: tag });
    }
});

// save the new tags (if any)
if (newTagsToSave.length > 0) {
    await Database.insert(newTagsToSave).into('tags');
}

// first, we have to get the IDs of all the incomingTags now that they have been saved
const allSavedPostTags = await Database.from('tags').whereIn('title', incomingTags);

// prepare post tags for insertion into pivot table
const postTagPivotEntries = [];
allSavedPostTags.forEach((tag) => {
    postTagPivotEntries.push({ post_id: postId, tag_id: tag.id });
});

// delete existing pivot entries
// then add all the new pivot entries
if (postTagPivotEntries.length > 0) {
  // delete existing post tag pivot entries for this post so we can start over
  const affectedRows = await Database.table('posts_tags').where({ post_id: postId }).delete();

  const updatedPostTagPivotEntries = await Database.insert(postTagPivotEntries).into('posts_tags');
}
```

My Post model has this:
```
tags() {
        return this.belongsToMany('App/Models/Tag', 'post_id', 'tag_id').pivotTable('posts_tags');
    }
```

and my Tag model has this:
```
posts() {
        return this.belongsToMany('App/Models/Post', 'tag_id', 'post_id').pivotTable('posts_tags');
    }
```

surely there's got to be a more elegant way?

**Comment by thetutlage** [comment]
So I imagine the UI being something like this

- Enter post details like (title, description)
- Then, there will be dropdown, which has a list of pre-created tags
- Maybe you also allow creating inline tags by just entering the tag title and hit enter 
- In either case, you will have the tag id

So when someone hits `Create Post`, you should be able to get

- postTitle
- description
- An array of tags ids

And this is how I will handle it

```ts
const post = await Post.create({})
await post.tags().sync(tagsIds)
```

In case of an update

```ts
const post = await Post.findOrFail(postId)
await post.tags().sync(tagsIds)
```

**Comment by andyford** [comment]
Okay, I think [this tutorial](https://www.youtube.com/watch?v=NUfpIDMqx2k) and the associated [repo branch](https://github.com/patrickodacre/simple-api-adonis-js/tree/lucid_many_to_many_complete) go a long way to helping me understand the whole concept

Also note this comment =)
<img width="823" alt="Screen Shot 2020-07-20 at 14 38 26" src="https://user-images.githubusercontent.com/49811/87938449-e8ed6a80-ca96-11ea-9543-2d266f8f7e82.png">

**Reply by dink** [reply]
Thanks for going down the rabbit hole. I think this is a pretty solid example I can start implementing now.

**Comment by dink** [comment, marked as answer]
@andyford 

I'm still struggling with implementing the post tags. Do you mind sharing the relevant parts of your model, migration, and controller (perhaps view as well) so I can make sense of it all?

**Reply by andyford** [reply]
Hi @dink 

just as a reminder, I'm still using Adonis 4.1 ... things might be different in 5.x

My Post model has a tags relationship

```
tags() {
    return this.belongsToMany('App/Models/Tag').withTimestamps();
}
```

...and the only thing my Tag model has is the posts relation
```
class Tag extends Model {
    posts() {
        return this.belongsToMany('App/Models/Post').withTimestamps();
    }
}
```

My initial tag migration is like this:

```
'use strict';

/** @type {import('@adonisjs/lucid/src/Schema')} */
const Schema = use('Schema');

class TagSchema extends Schema {
    up() {
        this.create('tags', (table) => {
            table.increments();
            table.timestamps();
            table.text('title').notNullable().unique();
        });
    }

    down() {
        this.drop('tags');
    }
}

module.exports = TagSchema;
```

and the migration for the `post_tag` pivot table is named `1595024107044_post_tag_schema.js`
...if I recall correctly, Adonis is very particular about which entity comes first in the naming of either the migration file, the table name or both (I don't exactly remember anymore). But I do remember really stumbling over the setup of all of this

```
'use strict';

/** @type {import('@adonisjs/lucid/src/Schema')} */
const Schema = use('Schema');

class PostTagSchema extends Schema {
    up() {
        this.create('post_tag', (table) => {
            table.increments();
            table.timestamps();

            table.integer('post_id').unsigned().index('post_id');
            table.foreign('post_id').references('posts.id').onDelete('cascade');

            table.integer('tag_id').unsigned().index('tag_id');
            table.foreign('tag_id').references('tags.id').onDelete('cascade');
        });
    }

    down() {
        this.drop('post_tag');
    }
}

module.exports = PostTagSchema;
```

regarding the view, I'm using https://github.com/yairEO/tagify for the tag entry UI and the controller receives the tags as a stringified json array like so
`'[{"value":"cat"},{"value":"animal"},{"value":"mammal"},{"value":"Lorem Ipsum"}]'`

I'm probably doing WAY more than I need to but here is the relevant controller logic:

`const { title, description, tags, post_id } = request.all();`

for better or worse, my controller method handles creation of new posts or editing of existing posts, so i have this code:

```
let post;
if (post_id) {
    post = await Post.find(post_id);
} else {
    post = new Post();
}
```

then the handling of the tags (which again is probably total overkill)...

```
const parsedTags = JSON.parse(tags);
        const incomingTagTitles = [];

        parsedTags &&
            parsedTags.forEach((tag) => {
                let tagName = tag.value;

                // a lot of formatting which you may not need
                tagName = tagName.trim().toLowerCase();
                tagName = tagName.replace(/_/g, ' ');
                tagName = tagName.replace(/#/g, '');
                tagName = tagName.replace(/  +/g, ' ');
                tagName = tagName.replace(/\s\s+/g, ' ');
                tagName = emojiStrip(tagName);

                if (tagName !== '') {
                    incomingTagTitles.push(tagName.trim());
                }
            });

        // get any pre-existing tags from the list of incoming tags
        // then put their string titles into an array (existingTagTitles)
        const existingTags = await Database.from('tags').whereIn('title', incomingTagTitles);
        const existingTagTitles = existingTags.map((tag) => tag.title);

        // we only want to save the tags that weren't previously saved
        // so check incoming tags against previously saved tags and create array of new tags to save
        const newTagsToSave = incomingTagTitles
            .filter((tag) => !existingTagTitles.includes(tag))
            .map((tag) => ({ title: tag }));

        if (newTagsToSave.length > 0) {
            await post.tags().createMany(newTagsToSave);
        }

        // get the IDs of all the incoming tags
        // prepare post tags for insertion into pivot table
        const allSavedPostTags = await Database.from('tags').whereIn('title', incomingTagTitles);
        const postTagPivotIds = allSavedPostTags.map((tag) => tag.id);

        if (postTagPivotIds.length > 0) {
            await post.tags().sync(postTagPivotIds);
        }
```

**Reply by andyford** [reply]
also I highly  recommend this video. It helped me a lot https://www.youtube.com/watch?v=NUfpIDMqx2k&ab_channel=BuildtoLearn

**Reply by dink** [reply]
@andyford Thank you! I'm working through it but am encountering a few hiccups along the way -- probably overlooking something but too tired to think straight right now. Will report back. 

TBH, this is a lot more complicated than I initially assumed.

**Reply by dink** [reply]
@andyford 

**Update:** I've got it working! Need to iron out a few kinks, but generally everything's fine. Thanks again for all your help.
